### PR TITLE
add test for shape validation

### PR DESCRIPTION
**Pull request by jbkyang-nvi**
Tests: https://github.com/triton-inference-server/core/pull/350

### CODE REVIEW

File: qa/L0_input_validation/test.sh [hunk 1]
```diff
@@ -64,14 +65,22 @@ set +e
...
-    echo -e "\n***\n*** python_unittest.py FAILED. \n***"
+    echo -e "\n***\n*** input_validation_test.py FAILED. \n***"
     RET=1
 fi
 set -e
 
 kill $SERVER_PID
 wait $SERVER_PID
 
+pip install torch
```
Comment (GuanLuo) [inline qa/L0_input_validation/test.sh:76]: Why pytorch? Should be more flexible using Python backend

File: qa/L0_input_validation/input_shape_validation_test.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import pytest
+import torch
+from tritonclient.grpc.aio import InferenceServerClient, InferInput
+from tritonclient.utils import np_to_triton_dtype
+
+GRPC_PORT = 9653
+FIXED_LAST_DIM = 8
+
+
+def repo_dir():
```
Comment (GuanLuo) [inline qa/L0_input_validation/input_shape_validation_test.py:44]: Like the idea of moving all logic within Python, can this be common utils?

File: qa/L0_input_validation/input_shape_validation_test.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            )  # should fail here
+            t2 = asyncio.create_task(bob.infer("pt_identity", inputs_2))
+
+        # alice_result, bob_result = await asyncio.gather(t1, t2)
+        # print(f"{alice_result.as_numpy('OUTPUT0')=}")
+        # print(f"{bob_result.as_numpy('OUTPUT0')=}")
+        # server.terminate()
+        # assert np.allclose(
+        #     bob_result.as_numpy("OUTPUT0"), input_data_2
+        # ), "Bob's result should be the same as input"
```
Comment (GuanLuo) [inline qa/L0_input_validation/input_shape_validation_test.py:158]: Uncomment?

File: qa/L0_input_validation/input_shape_validation_test.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+async def server_terminated(client: InferenceServerClient, server_proc):
+    if server_proc is not None and (ret_code := server_proc.poll()) is not None:
+        _, stderr = server_proc.communicate()
+        print(stderr)
+        raise Exception(f"Tritonserver died with return code {ret_code}")
+
+
+@pytest.mark.asyncio
+async def test_shape_overlapped(repo_dir: str):
```
Comment (GuanLuo) [inline qa/L0_input_validation/input_shape_validation_test.py:105]: Can you add summary on how the test operates?

File: qa/L0_input_validation/input_shape_validation_test.py (hunk 2 above)
Comment (yinggeh) [reply qa/L0_input_validation/input_shape_validation_test.py:44]: Agree but will push back for now. This function needs to be generic then we can refactor other tests.

File: qa/L0_input_validation/test.sh (hunk 1 above)
Comment (jbkyang-nvi) [reply qa/L0_input_validation/test.sh:76]: This is taken from [this reproducer](https://github.com/triton-inference-server/server/issues/7171#issuecomment-2099531946)

File: qa/L0_input_validation/input_shape_validation_test.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import asyncio
+from pathlib import Path
+from subprocess import Popen
+from tempfile import TemporaryDirectory
+from typing import Optional
```
[the remark this answers is not in the document]
Comment (jbkyang-nvi) [reply qa/L0_input_validation/input_shape_validation_test.py:32]: please delete use of optional here

File: qa/L0_input_validation/input_shape_validation_test.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        ]
+        inputs_2[0].set_data_from_numpy(input_data_2)
+        with pytest.raises(Exception) as e_info:
+            server_terminated(
+                InferenceServerClient("localhost:" + str(GRPC_PORT)), server
+            )
+            t1 = asyncio.create_task(
+                alice.infer("pt_identity", inputs_1)
+            )  # should fail here
+            t2 = asyncio.create_task(bob.infer("pt_identity", inputs_2))
```
[the remark this answers is not in the document]
Comment (jbkyang-nvi) [reply qa/L0_input_validation/input_shape_validation_test.py:153]: remove t2 and just call `create_task` here

File: qa/L0_input_validation/input_shape_validation_test.py (hunk 3 above)
Comment (jbkyang-nvi) [reply qa/L0_input_validation/input_shape_validation_test.py:161]: remove?

**Review by yinggeh — APPROVED** [review]

**Pull request merged**